### ISSUE TITLE
[Request] "Translator mode" toggle for maintainers

### ISSUE THREAD

**Issue by chimericdream** [i18next/i18next#2251]
## 🚀 Feature Proposal

It would be awesome if we could toggle a "dev mode" or "translator mode" where i18next uses the default language, but instead of displaying translations, it displays the fully qualified keys of translated strings.

## Motivation

The company I work at has multiple teams contributing to the translations for our app, and the easiest way to collaborate is using the translation keys. However, on some screens, such as FAQ pages, the translations might be a list of objects. The problem is that when we switch to the dev language to get the keys, these arrays either cause an error (because our code expects arrays, not strings) or just display the key of the overall array. Because of this, we can't simply use `cimode` or an empty lang file with keys displayed.

### Workaround

For the moment, I am using a script to generate a new language. It makes a complete copy of all of the JSON files in our `locales/en` directory, then recursively replaces all of the values with the full key to that particular translation. In other words, it takes a file like this:

```json
{
    "pages": {
        "home": {
            "heading": "Home page",
            "subheading": "Woo hoo"
        },
        "faqs": {
            "heading": "Frequently asked questions",
            "entries": [
                {
                    "question": "This is a question",
                    "answer": "this is an answer"
                },
                {
                    "question": "This is another question",
                    "answer": "this is another answer"
                }
            ]
        }
    }
}
```

And turns it into this:

```json
{
    "pages": {
        "home": {
            "heading": "pages.home.heading",
            "subheading": "pages.home.subheading"
        },
        "faqs": {
            "heading": "pages.faqs.heading",
            "entries": [
                {
                    "question": "pages.faqs.entries[0].question",
                    "answer": "pages.faqs.entries[0].answer"
                },
                {
                    "question": "pages.faqs.entries[1].question",
                    "answer": "pages.faqs.entries[1].answer"
                }
            ]
        }
    }
}
```

The feature request I have in mind would essentially do this replacement on the fly instead of requiring a separate locale JSON to be generated and maintained.

## Example

Here are some screenshots of a sample help page with some FAQs.

English (dev/translator mode off)
![Screenshot 2024-11-19 at 9 41 52 AM](https://github.com/user-attachments/assets/ec5f4f30-3ef1-4b5b-add8-b6c8b172e361)

English (dev/translator mode on)
![Screenshot 2024-11-19 at 9 52 33 AM](https://github.com/user-attachments/assets/f92b6e3a-6ccc-4b5b-9f17-8539c249e4a5)

**Comment by adrai** [comment]
Can you please share some code? I don't really understand what you're asking for.

For example: these two t calls, returns an object as expected

```js
const i18next = require('i18next')
i18next.init({
  fallbackLng: 'en',
  resources: {
    en: {
      translation: {
        key: {
          with: {
            obj: 'a',
            or: {
              arr: ['asdf', 'ffff']
            }
          }
        }
      }
    },
    dev: {
      translation: {
        key: {
          with: {
            obj: 'DEV: a',
            or: {
              arr: ['asdf', 'ffff']
            }
          }
        }
      }
    }
  }
})
console.log(i18next.t('key', { lng: 'dev', returnObjects: true }))
console.log(i18next.t('key', { lng: 'en', returnObjects: true }))
```

**Comment by jamuhl** [comment]
No.  dev language more relates to what ever your devs put as default values -> denglish for english as an example (not reviewed english)

lng=cimode is what shows keys

**Comment by chimericdream** [comment]
Got it. In that case, my request stands, but applies to the `cimode` language. If I set the language in my app to `cimode` and try to load a key with `returnObjects: true`, I get just the key name itself, without any way of knowing the structure of the object or array is expected.

**Comment by jamuhl** [comment]
cimode does not know about other languages...it returns early providing the key you passed for unit tests -> so there is no way for it to magically know your structure...

you might use your dev language with a hardcoded defaultValue

**Comment by adrai** [comment]
btw: these two should also return the same fallback lng value

```js
const i18next = require('i18next')
i18next.init({
  fallbackLng: 'en',
  resources: {
    en: {
      translation: {
        key: {
          with: {
            obj: 'a',
            or: {
              arr: ['asdf', 'ffff']
            }
          }
        }
      }
    }
  }
})
console.log(i18next.t('key', { lng: 'dev', returnObjects: true }))
console.log(i18next.t('key', { lng: 'en', returnObjects: true }))
```

**Comment by chimericdream** [comment]
For this code:

```
const fake = i18n.createInstance();
await fake.init({
    fallbackLng: 'en',
    resources: {
        en: {
            translation: {
                key: {
                    with: {
                        obj: 'a',
                        or: {
                            arr: ['asdf', 'ffff'],
                        },
                    },
                },
            },
        },
    },
});

console.log(fake.t('key', { returnObjects: true }));
console.log(fake.t('key', { lng: 'cimode', returnObjects: true }));
```

This is the console output:

![Screenshot 2024-11-19 at 11 03 34 AM](https://github.com/user-attachments/assets/d0ba9097-6ac7-4b89-a629-1836e76c3843)

What I'd like is a way to get output like this. (faked for the purpose of the screenshot)

![Screenshot 2024-11-19 at 11 05 55 AM](https://github.com/user-attachments/assets/7d6cb51c-6cae-4271-8df4-5b01472b1dbd)

If I'm interpreting things correctly, it sounds like the recommended way to do this would be to create a custom language (e.g. `dev`) as a clone of the default language, but with all the values replaced by their associated key. Is that accurate?

**Comment by chimericdream** [comment]
The core motivation behind this is that we want our translators to be able to say something along the lines of "Update the Spanish translation for `pages.faq.items[3].line2` to the following: ...". There is no current mechanism to get the fully resolved `pages.faq.items[3].line2` key to display when using `returnObjects: true`, because `cimode` will just return the string `pages.faq.items`.

**Comment by adrai** [comment]
> If I'm interpreting things correctly, it sounds like the recommended way to do this would be to create a custom language (e.g. dev) as a clone of the default language, but with all the values replaced by their associated key. Is that accurate?

Yes, there is no such magic functionality right now...

**Comment by chimericdream** [comment]
> > If I'm interpreting things correctly, it sounds like the recommended way to do this would be to create a custom language (e.g. dev) as a clone of the default language, but with all the values replaced by their associated key. Is that accurate?
> 
> Yes, there is no such magic functionality right now...

Got it. I'll tweak the feature request but otherwise leave it open. I can write a script to work around it for now, but it would be great to be able to toggle some kind of "translator mode" that would use the default language but display keys instead of values.

**Comment by chimericdream** [comment]
> I‘m rather sure this won‘t be added. A custom post processor should do the trick https://www.i18next.com/misc/creating-own-plugins post processor

Yeah. This is basically what I am doing, albeit with a plain script instead of a post-processor.

**Comment by adrai** [comment]
stale... so closing for now...
